### PR TITLE
Fix a file coping to itself

### LINKED ISSUE #3155: Copy a file to itself will remove the file
If a file is copied to itself, it will remove the file and state that the file does not exist. The expected result should (probably) be that the file is preserved as is.

```elixir
iex(1)> File.ls!
[]
iex(2)> File.write!("test_file", ":)")
:ok
iex(3)> File.ls!
["test_file"]
iex(4)> File.cp("test_file", "test_file")
{:error, :enoent}
iex(5)> File.ls!
[]
```

### PR DESCRIPTION
**Pull request by lexmag**
Closes #3155.
Cancels #3156.

### CODE REVIEW

**Comment by lexmag** [comment]
@josevalim this is ready for review.

File: lib/elixir/lib/file.ex [hunk 1]
```diff
@@ -461,7 +461,7 @@ defmodule File do
   error will be returned.
   """
   @spec cp(Path.t, Path.t, (Path.t, Path.t -> boolean)) :: :ok | {:error, posix}
-  def cp(source, destination, callback \\ fn(_, _) -> true end) do
+  def cp(source, destination, callback \\ &path_differs?/2) do
```
Comment (josevalim) [inline lib/elixir/lib/file.ex]: This means if the user is passing a custom function, they will need to add this logic himself. I don't think this is behaviour we want, right? We don't want to ever copy, we don't want to even invoke the callback. No?

Comment (lexmag) [inline lib/elixir/lib/file.ex]: It feels that `path_differs?/2` check out of callback makes this function more rigid.
My thoughts are to let user decide how to handle existence of the destination file even if it's the source itself. I think we could have a public alternative to `path_differs?/2`. However, it's hard to say what the best for a user before we see it in use. :bowtie:

Comment (josevalim) [inline lib/elixir/lib/file.ex]: Given that returning true would create an empty file, which is a bug, i
really don't think we want to allow the user to have this option. :)

Comment (lexmag) [inline lib/elixir/lib/file.ex]: `path_differs?/2` is moved out of callback.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**